src/System.Text.Json/src/System/Text/Json/JsonUtf8Reader.cs

```diff
@@ -0,0 +1,1274 @@
+    /// Since this type is a ref struct, it does not directly support async. However, it does provide
+    /// support for reentrancy to read incomplete data, and continue reading once more data is presented.
+    /// To be able to set max depth while reading OR allow skipping comments, create an instance of 
+    /// <see cref="JsonReaderState"/> and pass that in to the reader.
+    /// </summary>
+    public ref partial struct JsonUtf8Reader
+    {
+        private ReadOnlySpan<byte> _buffer;
+
+        private bool _isFinalBlock;
+
+        private ulong _stackFreeContainer;
+        private long _lineNumber;
+        private long _lineBytePosition;
+        private int _consumed;
+        private int _currentDepth;
+        private int _maxDepth;
+        private bool _inObject;
+        private bool _isNotPrimitive;
+        private JsonTokenType _tokenType;
+        private JsonReaderOptions _readerOptions;
+        private Stack<JsonTokenType> _stack;
+
+        private long _totalConsumed;
+        private bool _isLastSegment;
+        private readonly bool _isSingleSegment;
+        private SequencePosition _currentPosition;
+        //private SequencePosition _nextPosition; //TODO: Required for reading JSON from a ReadOnlySequeunce.
+        private ReadOnlySequence<byte> _sequence;
+
+        private bool IsLastSpan => _isFinalBlock && (_isSingleSegment || _isLastSegment);
+
+        /// <summary>
+        /// Lets the caller know which of the two 'Value' properties to read to get the 
+        /// token value. For input data within a ReadOnlySequence&lt;byte&gt; this will
+        /// always return false. For input data within a ReadOnlySequence&lt;byte&gt;, this
+        /// will only return true if the token straddles more than a single segment and
+        /// hence couldn't be represented as a span.
+        /// </summary>
+        public bool IsValueMultiSegment { get; private set; }
+
+        /// <summary>
+        /// Gets the value of the last processed token as a ReadOnlySpan&lt;byte&gt; slice
+        /// of the input payload. If the JSON is provided within a ReadOnlySequence&lt;byte&gt;
+        /// and the slice that represents the token value fits in a single segment, then
+        /// ValueSpan will contain the sliced value since it can be represented as a span.
+        /// </summary>
+        public ReadOnlySpan<byte> ValueSpan { get; private set; }
+
+        /// <summary>
+        /// Gets the value of the last processed token as a ReadOnlySequence&lt;byte&gt; slice
+        /// of the input payload. If the JSON is provided within a ReadOnlySpan&lt;byte&gt;
+        /// this will always return an empty sequence. Also, if the JSON is provided within a
+        /// ReadOnlySequence&lt;byte&gt; and the slice that represents the token value fits in
+        /// a single segment, then ValueSpan will contain the sliced value since it can be
+        /// represented as a span.
+        /// </summary>
```

Comment:
The "Also, ..." sentence either means A) this property returns the empty sequence for a ROSequence when ValueSpan would be sensible or B) everything still works (aka the sentence is a distraction, not a summary).

If it's B, just remove the sentence altogether (or move it to `<remarks>` as a "in this case both properties are valid").  If it's A, the Span vs Sequence stuff seems better written as "if `IsValueMultiSegment` is `false` then the empty sequence is returned".

I'd also add a `<seealso cref="ValueSpan" />` (and reflexively from ValueSpan to here).